### PR TITLE
Elm 0.16 (new formula)

### PR DESCRIPTION
**Pull request by mietek**
Based on the Elm Platform [installation instructions](https://github.com/elm-lang/elm-platform#build-from-source), and the [installation script](https://raw.githubusercontent.com/elm-lang/elm-platform/master/installers/BuildFromSource.hs).

### CODE REVIEW

File: Library/Formula/elm.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  depends_on "cabal-install" => :build
+
+  setup_ghc_compilers
+
+  def install
+    compiler_path = buildpath/"elm-compiler"
+    mktemp do
+      mv Dir["#{buildpath}/*"], pwd
+      mv pwd, compiler_path
+    end
```
Comment (bfontaine) [inline Library/Formula/elm.rb]: Why not `(buildpath/"elm-compiler").install Dir["#{buildpath}/*"]` instead?

File: Library/Formula/elm.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  end
+
+  test do
+    src_path = testpath/"Hello.elm"
+    src_path.write <<-EOS.undent
+      import Html exposing (text)
+      main = text "Hello, world!"
+    EOS
+
+    system "elm", "package", "install", "evancz/elm-html", "--yes"
```
Comment (bfontaine) [inline Library/Formula/elm.rb]: Use `"#{bin}/elm"` instead of `"elm"` here.

File: Library/Formula/elm.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    out_path = testpath/"index.html"
+    system "elm", "make", src_path, "--output=#{out_path}"
```
Comment (bfontaine) [inline Library/Formula/elm.rb]: Same here.

File: Library/Formula/elm.rb (hunk 1 above)
Comment (mietek) [inline Library/Formula/elm.rb]: After the tarball is unpacked, we have the following directory structure inside `buildpath`:
```
file1
file2
...
fileN
```

For the next step of the build process, we want to have the following directory structure:
```
elm-compiler/file1
elm-compiler/file2
...
elm-compiler/fileN
```

Is there a way to ask Homebrew to unpack the tarball in a subdirectory?

File: Library/Formula/elm.rb (hunk 2 above)
Comment (mietek) [inline Library/Formula/elm.rb]: OK.

File: Library/Formula/elm.rb (hunk 3 above)
Comment (mietek) [inline Library/Formula/elm.rb]: OK.

File: Library/Formula/elm.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Formula/elm.rb]: > Is there a way to ask Homebrew to unpack the elm-compiler tarball in a subdirectory of the buildpath?

Unfortunately not; can you explain why that's required?

Comment (mietek) [inline Library/Formula/elm.rb]: Ultimately, we want to have the following directory structure inside the buildpath:
```
elm-compiler/...
elm-package/...
elm-make/...
elm-repl/...
elm-reactor/...
```
Each of these subdirectories contains one Cabal package.  In the next step of the build process, we add each of these packages to a single Cabal sandbox.  This allows Cabal to find the correct version of these packages, and to successfully resolve dependencies.

Being able to ask Homebrew to unpack in a subdirectory would be nice, but it’s not required, as the formula is correct, if slightly inelegant.

**Comment by mietek** [comment]
Is there anything left to do here?

**Comment by bfontaine** [comment]
@mietek You should be able to replace the `mktemp` block thing with `(buildpath/"elm-compiler").install Dir["*"]`.

**Comment by mietek** [comment]
OK, done.

**Comment by bfontaine** [comment]
Thank you for your contribution to Homebrew @mietek; we appreciate it! :tada:

**Pull request closed without merging**